micropython/usb/usb-device/usb/device/core.py

```diff
@@ -308,12 +320,26 @@ def _submit_xfer(self, ep_addr, data, done_cb=None):
         self._ep_cbs[ep_addr] = done_cb or True
         return self._usbd.submit_xfer(ep_addr, data)
 
+    def _xfer_pending(self, ep_addr):
+        # Singleton function to return True if transfer is pending on this endpoint.
+        #
+        # Generally, drivers should call Interface.xfer_pending() instead. See that
+        # function for more documentation.
+        return bool(self._ep_cbs[ep_addr]) or (
+            self._cb_ep == ep_addr and self._cb_thread != get_ident()
+        )
+
     def _xfer_cb(self, ep_addr, result, xferred_bytes):
         # Singleton callback from TinyUSB custom class driver when a transfer completes.
         cb = self._ep_cbs.get(ep_addr, None)
+        self._cb_thread = get_ident()
+        self._cb_ep = ep_addr  # Track while callback is running
         self._ep_cbs[ep_addr] = None
-        if callable(cb):
-            cb(ep_addr, result, xferred_bytes)
+        try:
+            if callable(cb):
```

Comment:
The value of `cb` should be `True` if a transfer is pending but no callback is set (so it can be tested for truthiness elsewhere), or will be the callback function otherwise.

Have put in a comment to make this clearer.

(As an aside, this is the kind of thing that typing annotations really help with!)

Reply:
Won't the `cb` be `None` in some cases, due to the `get(ep_addr, None)` above?

I was also thinking that the try/finally (which is expensive) only needs to be done if the cb is going to be called.

Reply:
> Won't the cb be None in some cases, due to the get(ep_addr, None) above?

If it's None here then it's some kind of semi-error case because TinyUSB is giving us a callback for a transfer we didn't initiate. It might happen if the device configuration is being changed, but it's not something which happens normally..

> I was also thinking that the try/finally (which is expensive) only needs to be done if the cb is going to be called.

Oh,right. I guess there's a version of this which is:

```
if callable(cb):
    try: 
        ...
```

I think 90% of cases (and especially performance critical cases) there will be a callback here, though.

I'll update the comment again, at least.

Reply:
(Updated, realised moving the `if` outside the `try...finally` block is fiddly because we always want the code in the finally to run.)

Reply:
I guess you could also do `if not callable(cb): return` near the start of this function, to avoid most of the work if there's no callback.  But if the callback is there in a large proportion of cases, then the way it is now is fine.